### PR TITLE
Fixing semantics borders on mobile web

### PR DESCRIPTION
**Pull request by nturgut**
Changing the css matrix transformation to only add 2d styling values (top, left, width, height). Calculating bounding rectangle for the transform so that if the semantics boundary is not an ordinary 2d translate, still focus is around the semantics area. This is similar to the Flutter native behaviour. 

Fixes: https://github.com/flutter/flutter/issues/68225

### CODE REVIEW

**Comment by nturgut** [comment]
@goderbauer can you confirm that the semantics focus node's border can be represented with 2d translation only? Or is it possible to receive  a matrix we can't represent with top/left.

I tested this transform on a native Flutter app:
`transform: Matrix4.skewY(0.3)..rotateZ(-pi / 12.0),` 

still the screen reader focus node's border was a rectangle (not a parallelogram)

File: lib/web_ui/lib/src/engine/util.dart [hunk 1]
```diff
@@ -86,6 +91,19 @@ String float64ListToCssTransform(Float32List matrix) {
...
+/// Converts [matrix] to coordinates on 2 dimensional space.
+///
+/// Only used by semantics node styling calculations, since semantics
+/// node does not receive 3D transforms, only tx and ty values of the matrix
+/// is used.
+Position2D float64ListToCss2dPositionSemantics(Float32List matrix) {
+  assert(transformKindOf(matrix) != TransformKind.transform2d);
+  assert(matrix.length == 16);
+  final Float32List m = matrix;
+  print('top left: ${matrix[13]} ${matrix[12]} semantics');
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/util.dart:103]: left over print

File: lib/web_ui/lib/src/engine/semantics/semantics.dart [hunk 2]
```diff
@@ -882,9 +882,11 @@ class SemanticsObject {
     }
 
     if (!effectiveTransformIsIdentity) {
+      final Position2D position2d =
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/semantics/semantics.dart:885]: suggestion: translate(0, 0) and use the resulting offset, instead of adding Position2D and adding matrix utils.

**Comment by goderbauer** [comment]
> @goderbauer can you confirm that the semantics focus node's border can be represented with 2d translation only? Or is it possible to receive a matrix we can't represent with top/left.

It's really up to the embedder to decide how to and where to draw the focus rect. The flutter framework provides you with a `Rect` and an arbitrary transform `Matrix4` for each semantics node. It's upto the embedder to translate that into a focus rect. As far as I know, most (all?) platforms we support just draw the focus rect as the bounding box around the transformed rect, see Android for example: https://github.com/flutter/engine/blob/4f89ff7a9c979ccab47c48524886ecee477f156f/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L2266

However, in theory, if the platform supports it, you could draw a parallelogram as a focus rect. Flutter provides you all the data for this. Example:

```
Align(
      alignment: Alignment.topLeft,
      child: Transform(
        transform: Matrix4.skewY(0.3)..rotateZ(-pi / 12.0),
        child: Semantics(
          container: true,
          label: 'hello',
          child: Container(
            height: 100,
            width: 100,
            color: Colors.green,
          ),
        ),
      )
    );
```

This creates the following semantics node:

```
I/flutter (18226):        └─SemanticsNode#4
I/flutter (18226):            Rect.fromLTRB(0.0, 0.0, 100.0, 100.0) with transform
I/flutter (18226):            [0.9659258262890683,0.25881904510252074,0.0,0.0;
I/flutter (18226):            0.039976827402816106,1.045987939028626,0.0,0.0; 0.0,0.0,1.0,0.0;
I/flutter (18226):            0.0,0.0,0.0,1.0]
I/flutter (18226):            label: "hello"
I/flutter (18226):            textDirection: ltr
```

and if you apply that matrix to the rect you should get a parallelogram.

I believe just reading the translation values of the transform matrix is not enough, though. In the example above, if you replace the transform parameter with something like `Matrix4.identity()..scale(2.0)` wouldn't you draw a focus rect that's too small since it's missing the scale? Or is that applied elsewhere in web land?

**Comment by nturgut** [comment]
> > @goderbauer can you confirm that the semantics focus node's border can be represented with 2d translation only? Or is it possible to receive a matrix we can't represent with top/left.
> 
> It's really up to the embedder to decide how to and where to draw the focus rect. The flutter framework provides you with a `Rect` and an arbitrary transform `Matrix4` for each semantics node. It's upto the embedder to translate that into a focus rect. As far as I know, most (all?) platforms we support just draw the focus rect as the bounding box around the transformed rect, see Android for example:
> 
> https://github.com/flutter/engine/blob/4f89ff7a9c979ccab47c48524886ecee477f156f/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L2266
> 
> However, in theory, if the platform supports it, you could draw a parallelogram as a focus rect. Flutter provides you all the data for this. Example:
> 
> ```
> Align(
>       alignment: Alignment.topLeft,
>       child: Transform(
>         transform: Matrix4.skewY(0.3)..rotateZ(-pi / 12.0),
>         child: Semantics(
>           container: true,
>           label: 'hello',
>           child: Container(
>             height: 100,
>             width: 100,
>             color: Colors.green,
>           ),
>         ),
>       )
>     );
> ```
> 
> This creates the following semantics node:
> 
> ```
> I/flutter (18226):        └─SemanticsNode#4
> I/flutter (18226):            Rect.fromLTRB(0.0, 0.0, 100.0, 100.0) with transform
> I/flutter (18226):            [0.9659258262890683,0.25881904510252074,0.0,0.0;
> I/flutter (18226):            0.039976827402816106,1.045987939028626,0.0,0.0; 0.0,0.0,1.0,0.0;
> I/flutter (18226):            0.0,0.0,0.0,1.0]
> I/flutter (18226):            label: "hello"
> I/flutter (18226):            textDirection: ltr
> ```
> 
> and if you apply that matrix to the rect you should get a parallelogram.
> 
> I believe just reading the translation values of the transform matrix is not enough, though. In the example above, if you replace the transform parameter with something like `Matrix4.identity()..scale(2.0)` wouldn't you draw a focus rect that's too small since it's missing the scale? Or is that applied elsewhere in web land?

Thanks a lot for the insight. I can see that we are loosing the scale information here, so that's important to address.

Let me rephrase what I understand from the parallelogram example.  Other embedder's still does not have a parallelogram shaped focus node. The focus node is a rectangle bounding the parallelogram. I think this is also another area to address on the PR.

File: lib/web_ui/lib/src/engine/semantics/semantics.dart (hunk 2 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/semantics/semantics.dart:885]: I'm not sure how `translate(0, 0)` helps here. But could the PR be simplified to this?
```dart
final Vector3 translation = effectiveTransform.getTranslation();
element.style
  ..top = '${translation.y}px'
  ..left = '${translation.x}px';
```
or:
```dart
final Matrix4 translationOnly =
    Matrix4.identity()
    ..setTranslation(effectiveTransform.getTranslation());
element.style
    ..transformOrigin = '0 0 0'
    ..transform = matrix4ToCssTransform(translationOnly);
```

Comment (nturgut) [reply lib/web_ui/lib/src/engine/semantics/semantics.dart:885]: Both options does not work unfortunately.
a) Boundary does not get correctly recognized by the screen reader
```
..transformOrigin = '0 0 0'
    ..transform = matrix4ToCssTransform(translationOnly);
```
b) Loses other information.
```
element.style
  ..top = '${translation.y}px'
  ..left = '${translation.x}px';
```

I'll probably still go with option 2 though, since, transforms in 3D are a rare case and now most of the common cases are not working as intended (shown in the issue).

Comment (nturgut) [reply lib/web_ui/lib/src/engine/semantics/semantics.dart]: @ferhatb I used the method you sent yesterday for computing the bounding rectangle. Thanks!

**Review by ferhatb — APPROVED** [review]
